### PR TITLE
Fix | Adding type convertor support for SqlConnectionEncryptOption

### LINKED ISSUE #1894: SQLConnectionStringBuilder 5.0 is not Backwards Compatible with appSettings.json
### Describe the bug
With SqlClient 4, standard appSettings.json standard config file mapping from something like 
`
{
    "MyDb": {
        "Components": {
            "UserID": "blah",
            "Password": "foo",
            "DataSource": "hostname",
            "Encrypt": false
        }
    }
}`
to a SQLConnectionStringBuilder property 
worked. With SqlClient 5.0, the resulting connection string is now returning Encrypt=True

I've tried changing encrypt value in json to "false", "False", "SqlConnectionEncryptOption.Optional", and "Optional". All of those still result in connection string with Encrypt=True.

### To reproduce

```c#
    public class SqlConnectionString
    {
        public SqlConnectionStringBuilder Components { get; set; } = new();

        public override string ToString()
        {
            return Components.ConnectionString;
        }
    }

    public class Settings
    {
        [Required]
        public SqlConnectionString? MyDb { get; set; }
    }
```

```c#
            hostBuilder.ConfigureServices((ctx, services) =>
            {
                    var configuration = ctx.Configuration;
                    services.Configure<Settings>(configuration);
                    var settings = configuration.Get<Settings>();
                    Assert.Contains("Encrypt=false", settings.MyDb.ToString());
            }
```

### Expected behavior
Expecting appSettings.json (or environment variable override of it) setting of "Encrypt": false to result in Encrypt=false in resulting connection string.

### Further technical details
Microsoft.Data.SqlClient version: 5.0.1
.NET target: 6
SQL Server version: n/a
Operating system: Windows 10

**Additional context**
We're using the SqlConnectionStringBuilder to assemble/validate connection string components that are coming from different places (environment variables in helm chart, Azure keyvault, appSettings.json) into a single connection string. Our environment doesn't yet support encrypt=true/mandatory.

### PR DESCRIPTION
**Pull request by arellegue**
fixes #1894 
Added type converter for SqlConnectionEncryptOption.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+
```
Comment (JRahnama) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs:5]: remove the extra line here.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/Microsoft.Data.SqlClient.Tests.csproj [hunk 2]
```diff
@@ -66,6 +66,7 @@
     <Compile Include="..\..\src\Microsoft\Data\Common\MultipartIdentifier.cs" />
   </ItemGroup>
   <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Hosting" Version="6.0.0" />
```
Comment (JRahnama) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/Microsoft.Data.SqlClient.Tests.csproj:69]: do not add version directly here. You can add version number in version.props and use it from there. See other versions in the same file as a reference.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 3]
```diff
@@ -495,5 +526,61 @@ internal static void CheckEncryptType(SqlConnectionStringBuilder builder, SqlCon
             Assert.IsType<SqlConnectionEncryptOption>(builder.Encrypt);
             Assert.Equal(expectedValue, builder.Encrypt);
         }
+
+        internal void ExecuteConnectionStringFromJsonTests(string encryptOption, string result)
+        {
+            var settings = LoadSettingsFromJsonStream<UserDbConnectionStringSettings>(encryptOption);
```
Comment (JRahnama) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:532]: use explicit types instead of var.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 4]
```diff
@@ -495,5 +526,61 @@ internal static void CheckEncryptType(SqlConnectionStringBuilder builder, SqlCon
...
+            var connectionString = settings!.UserDb!.ToString();
+            Assert.Contains($"Encrypt={result}", connectionString, StringComparison.InvariantCultureIgnoreCase);
+        }
+
+        internal void ExecuteConnectionStringFromJsonThrowsException(string encryptOption)
+        {
+            Assert.Throws<System.InvalidOperationException>(() => LoadSettingsFromJsonStream<UserDbConnectionStringSettings>(encryptOption));
+        }
+
+        TSettings LoadSettingsFromJsonStream<TSettings>(string encryptOption) where TSettings : class
```
Comment (JRahnama) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:542]: this could be static.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        // conversion. Typically, this interface is used at design time to 
+        // provide information about the design-time container.
+        /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnectionEncryptOptionConverter.xml' path='docs/members[@name="SqlConnectionEncryptOptionConverter"]/SqlConnectionEncryptOptionConverter/CanConvertFrom/*'/>
+        public override bool CanConvertFrom(ITypeDescriptorContext context, Type sourceType)
+        {
+            if (sourceType == typeof(string))
+            {
+                return true;
+            }
+            return base.CanConvertFrom(context, sourceType);
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs:25]: I am not sure what you mean by increase code coverage.  This source code was taken from Microsoft example code.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            {
+                return SqlConnectionEncryptOption.Parse(value.ToString());
+            }
+            throw new Exception("Value to convert must be of string type!");
+        }
+        // Overrides the ConvertTo method of TypeConverter.
+        /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnectionEncryptOptionConverter.xml' path='docs/members[@name="SqlConnectionEncryptOptionConverter"]/SqlConnectionEncryptOptionConverter/ConvertTo/*'/>
+        public override object ConvertTo(ITypeDescriptorContext context, CultureInfo culture, object value, Type destinationType)
+        {
+            return base.ConvertTo(context, culture, value, destinationType);
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs:41]: This code was taken from Microsoft example. This function will never be used to convert from SqlConnectionEncryptOption into a string.  Therefore, I am deleting this override.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 7]
```diff
@@ -495,5 +526,61 @@ internal static void CheckEncryptType(SqlConnectionStringBuilder builder, SqlCon
             Assert.IsType<SqlConnectionEncryptOption>(builder.Encrypt);
             Assert.Equal(expectedValue, builder.Encrypt);
         }
+
+        static internal void ExecuteConnectionStringFromJsonTests(string encryptOption, string result)
+        {
+            UserDbConnectionStringSettings settings = LoadSettingsFromJsonStream<UserDbConnectionStringSettings>(encryptOption);
+            string connectionString = settings!.UserDb!.ToString();
+            Assert.Contains($"Encrypt={result}", connectionString, StringComparison.InvariantCultureIgnoreCase);
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:534]: Done.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs:24]: You should try to cover the code with unit tests as much as possible. If the current tests don't hit it, you should add a test in a way that accepts a none-string value.

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs]: In adding tests, the goal is not to assess a specific condition. You should focus on and cover all possible execution paths.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 8]
```diff
@@ -495,5 +540,56 @@ internal static void CheckEncryptType(SqlConnectionStringBuilder builder, SqlCon
...
+    internal class UserSqlConnectionString
+    {
+        public SqlConnectionStringBuilder UserComponents { get; set; } = new();
+
+        public override string ToString()
+        {
+            return UserComponents!.ConnectionString;
+        }
     }
+    #endregion
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:594]: You can move these lines to the same region with the tests, including the helper function.

File: tools/props/Versions.props [hunk 9]
```diff
@@ -74,6 +74,7 @@
     <MicrosoftSqlServerTypesVersionNet>160.1000.6</MicrosoftSqlServerTypesVersionNet>
     <BenchmarkDotNetVersion>0.13.2</BenchmarkDotNetVersion>
     <SystemServiceProcessServiceControllerVersion>6.0.0</SystemServiceProcessServiceControllerVersion>
+	<MicrosoftExtensionsHosting>6.0.0</MicrosoftExtensionsHosting>
```
Comment (DavoudEshtehari) [inline tools/props/Versions.props:77]:
```suggestion
    <MicrosoftExtensionsHosting>6.0.0</MicrosoftExtensionsHosting>
```

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs]: Added SqlConnectionEncryptOptionConverter unit tests

File: tools/props/Versions.props [hunk 10]
```diff
@@ -31,6 +31,7 @@
     <SystemBuffersVersion>4.5.1</SystemBuffersVersion>
     <SystemTextEncodingsWebVersion>6.0.0</SystemTextEncodingsWebVersion>
     <MicrosoftSourceLinkGitHubVersion>1.1.0</MicrosoftSourceLinkGitHubVersion>
+	<MicrosoftExtensionsHosting>6.0.0</MicrosoftExtensionsHosting>
```
[the remark this answers is not in the document]
Comment (arellegue) [reply tools/props/Versions.props]: It has been moved.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs (hunk 8 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs]: The helper classes have been moved to the same region with the tests.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 11]
```diff
@@ -495,5 +526,61 @@ internal static void CheckEncryptType(SqlConnectionStringBuilder builder, SqlCon
...
+        static internal void ExecuteConnectionStringFromJsonTests(string encryptOption, string result)
+        {
+            UserDbConnectionStringSettings settings = LoadSettingsFromJsonStream<UserDbConnectionStringSettings>(encryptOption);
+            string connectionString = settings!.UserDb!.ToString();
+            Assert.Contains($"Encrypt={result}", connectionString, StringComparison.InvariantCultureIgnoreCase);
+        }
+
+        static internal void ExecuteConnectionStringFromJsonThrowsException(string encryptOption)
+        {
+            Assert.Throws<System.InvalidOperationException>(() => LoadSettingsFromJsonStream<UserDbConnectionStringSettings>(encryptOption));
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs]: The expected exception content has been assessed too.

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 12]
```diff
@@ -4,6 +4,14 @@
...
 using System;
 using System.Collections.Generic;
+using System.ComponentModel;
+using System.ComponentModel.DataAnnotations;
+using System.IO;
+using System.Text;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Newtonsoft.Json.Linq;
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:14]:
```suggestion
```

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 13]
```diff
@@ -495,5 +624,68 @@ internal static void CheckEncryptType(SqlConnectionStringBuilder builder, SqlCon
...
+                .Build();
+
+            return settingsOut;
+        }
     }
+
+    #region ConnectionStringFromJsonRequiredClasses
+    // These 2 classes will be used by ConnectionStringFromJsonTests only
+
+    #endregion
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:690]:
```suggestion
```

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 14]
```diff
@@ -495,5 +624,68 @@ internal static void CheckEncryptType(SqlConnectionStringBuilder builder, SqlCon
...
+                    services.AddOptions();
+                    services.Configure<TSettings>(ctx.Configuration);
+                    settingsOut = configuration.Get<TSettings>();
+                })
+                .Build();
+
+            return settingsOut;
+        }
+
+        private static TSettings LoadSettingsFromBadJsonStream<TSettings>(int encryptOption) where TSettings : class
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:660]: Zero reference type! Have you forgotten to consume it or it's redundant?

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 15]
```diff
@@ -468,6 +476,127 @@ public void EncryptTryParseInvalidValuesReturnsFalse(string value)
...
+        internal class UserSqlConnectionString
+        {
+            public SqlConnectionStringBuilder UserComponents { get; set; } = new();
+
+            public override string ToString()
+            {
+                return UserComponents!.ConnectionString;
+            }
+        }
+        #endregion
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:598]:
```suggestion

        internal static void ExecuteConnectionStringFromJsonThrowsException(string encryptOption)
        {
            var exception = Assert.Throws<InvalidOperationException>(() => LoadSettingsFromJsonStream<UserDbConnectionStringSettings>(encryptOption));
            Assert.Contains("Failed to convert configuration", exception.Message, StringComparison.Ordinal);
        }

        private static TSettings LoadSettingsFromJsonStream<TSettings>(string encryptOption) where TSettings : class
        {
            TSettings settingsOut = null;

            Host.CreateDefaultBuilder()
                .ConfigureAppConfiguration((ctx, configBuilder) =>
                {
                    // Note: Inside string interpolation, a { should be {{ and a } should be }}
                    // First, declare a stringified JSON
                    var json = $"{{ \"UserDb\": {{ \"UserComponents\": {{ \"NetworkLibrary\": \"DBMSSOCN\", \"UserID\": \"user\", \"Password\": \"password\", \"DataSource\": \"localhost\", \"InitialCatalog\": \"catalog\", \"Encrypt\": \"{encryptOption}\" }}}}}}";
                    // Load the stringified JSON as a stream into the configuration builder
                    configBuilder.AddJsonStream(new MemoryStream(Encoding.ASCII.GetBytes(json)));
                    configBuilder.AddEnvironmentVariables();
                })
                .ConfigureServices((ctx, services) =>
                {
                    var configuration = ctx.Configuration;
                    services.AddOptions();
                    services.Configure<TSettings>(ctx.Configuration);
                    settingsOut = configuration.Get<TSettings>();
                })
                .Build();

            return settingsOut;
        }
        #endregion
```

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 16]
```diff
@@ -495,5 +624,68 @@ internal static void CheckEncryptType(SqlConnectionStringBuilder builder, SqlCon
...
+                {
+                    var configuration = ctx.Configuration;
+                    services.AddOptions();
+                    services.Configure<TSettings>(ctx.Configuration);
+                    settingsOut = configuration.Get<TSettings>();
+                })
+                .Build();
+
+            return settingsOut;
+        }
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs:658]: Moved to the same region.
```suggestion
```

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+        // Overrides the ConvertFrom method of TypeConverter.
+        /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnectionEncryptOptionConverter.xml' path='docs/members[@name="SqlConnectionEncryptOptionConverter"]/SqlConnectionEncryptOptionConverter/ConvertFrom/*'/>
+        public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, object value)
+        {
+            if (value is string)
+            {
+                return SqlConnectionEncryptOption.Parse(value.ToString());
+            }
+            throw new ArgumentException(StringsHelper.GetString(Strings.SqlConvert_ConvertFailed, value.GetType().FullName, typeof(SqlConnectionEncryptOption).FullName));
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs:45]: `AdapterUtil` is a helper class to manage exceptions.
```suggestion
            throw ADP.ConvertFailed(value.GetType(), typeof(SqlConnectionEncryptOption), null);
```

File: src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+        // Overrides the ConvertTo method of TypeConverter.
+        /// <include file='../../../../../../doc/snippets/Microsoft.Data.SqlClient/SqlConnectionEncryptOptionConverter.xml' path='docs/members[@name="SqlConnectionEncryptOptionConverter"]/SqlConnectionEncryptOptionConverter/ConvertTo/*'/>
+        public override object ConvertTo(ITypeDescriptorContext context, CultureInfo culture, object value, Type destinationType)
+        {
+            if (destinationType == typeof(string))
+            {
+                return base.ConvertTo(context, culture, value, destinationType);
+            }
+            throw new ArgumentException(StringsHelper.GetString(Strings.SqlConvert_ConvertFailed, value.GetType().FullName, typeof(SqlConnectionEncryptOption).FullName));
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionEncryptOptionConverter.cs:56]:
```suggestion
            throw ADP.ConvertFailed(value.GetType(), typeof(SqlConnectionEncryptOption), null);
```

**Review by DavoudEshtehari — COMMENTED** [review]
The xml files are only required for public APIs. In this case, if `SqlConnectionEncryptOptionConverter` type is going to be `internal`, remove the xml file and inline dependencies.

File: src/Microsoft.Data.SqlClient.sln [hunk 19]
```diff
@@ -121,6 +121,7 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Microsoft.Data.SqlClient",
 		..\doc\snippets\Microsoft.Data.SqlClient\SqlConnectionAttestationProtocol.xml = ..\doc\snippets\Microsoft.Data.SqlClient\SqlConnectionAttestationProtocol.xml
 		..\doc\snippets\Microsoft.Data.SqlClient\SqlConnectionColumnEncryptionSetting.xml = ..\doc\snippets\Microsoft.Data.SqlClient\SqlConnectionColumnEncryptionSetting.xml
 		..\doc\snippets\Microsoft.Data.SqlClient\SqlConnectionEncryptOption.xml = ..\doc\snippets\Microsoft.Data.SqlClient\SqlConnectionEncryptOption.xml
+		..\doc\snippets\Microsoft.Data.SqlClient\SqlConnectionEncryptOptionConverter.xml = ..\doc\snippets\Microsoft.Data.SqlClient\SqlConnectionEncryptOptionConverter.xml
```
Comment (JRahnama) [inline src/Microsoft.Data.SqlClient.sln:124]: is this still required?

Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient.sln:124]:
```suggestion
```

File: src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs [hunk 20]
```diff
@@ -468,6 +476,159 @@ public void EncryptTryParseInvalidValuesReturnsFalse(string value)
...
+
+        [Fact]
+        public void SqlConnectionEncryptOptionConverterCanConvertFromTest()
+        {
+            // Get a converter
+            SqlConnectionEncryptOption option = SqlConnectionEncryptOption.Parse("false");
+            TypeConverter converter = TypeDescriptor.GetConverter(option.GetType());
+            // Use the converter to determine if can convert from string data type
+            Assert.True(converter.CanConvertFrom(null, typeof(string)));
```
[the remark this answers is not in the document]
Comment (arellegue) [reply src/Microsoft.Data.SqlClient/tests/FunctionalTests/SqlConnectionStringBuilderTest.cs]: Additional optional message has been added.

File: src/Microsoft.Data.SqlClient.sln (hunk 19 above)
Comment (arellegue) [reply src/Microsoft.Data.SqlClient.sln]: The entry in the MDS solution for EncryptOptionConverter.xml has been deleted.

**Review by DavoudEshtehari — APPROVED** [review]

**Review by JRahnama — APPROVED** [review]
